File: apps/web/types/alert.ts
export interface AlertPolicyScope {
  monitor_ids?: string[];
  tags?: Record<string, string>;
}

export interface AlertConditions {
  consecutive_failures?: number;
  high_latency_ms?: number;
  packet_loss_percent?: number;
  ssl_expiring_days?: number;
  domain_expiring_days?: number;
  smtp_starttls_fail?: boolean;
  ntp_offset_ms?: number;
  dns_mismatch?: boolean;
}

export interface AlertPolicy {
  id: string;
  organization_id: string;
  project_id: string;
  name: string;
  scope: AlertPolicyScope;
  conditions: AlertConditions;
  severity: "info" | "warning" | "critical";
  opening_failures: number;
  resolving_successes: number;
  cooldown_seconds: number;
  renotify_seconds: number;
  enabled: boolean;
  channel_ids: string[];
  created_at: string;
  updated_at: string;
}

export type AlertState = "pending" | "firing" | "recovering" | "resolved" | "suppressed";

export interface Alert {
  id: string;
  organization_id: string;
  policy_id: string;
  monitor_id: string;
  state: AlertState;
  severity: "info" | "warning" | "critical";
  title: string;
  description: string;
  dedup_key: string;
  consecutive_failures: number;
  consecutive_successes: number;
  opened_at: string | null;
  resolved_at: string | null;
  created_at: string;
}

export interface NotificationChannel {
  id: string;
  organization_id: string;
  name: string;
  type: "email" | "webhook" | "telegram";
  config: Record<string, string>;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface AlertListResponse {
  items: Alert[];
}

export interface AlertPolicyListResponse {
  items: AlertPolicy[];
}

export interface NotificationChannelListResponse {
  items: NotificationChannel[];
}

File: packages/shared/domain/alert.go
package domain

import (
	"context"
	"time"
)

// NotificationChannel represents a destination for alert notifications.
type NotificationChannel struct {
	ID             string
	OrganizationID string
	WorkspaceID    *string
	Name           string
	Type           string
	Config         map[string]string
	Enabled        bool
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// AlertPolicy defines rules for when alerts should trigger.
type AlertPolicy struct {
	ID                 string
	OrganizationID     string
	WorkspaceID        *string
	Name               string
	Scope              AlertPolicyScope
	Conditions         AlertConditions
	Severity           string
	OpeningFailures    int
	ResolvingSuccesses int
	CooldownSeconds    int
	RenotifySeconds    int
	Enabled            bool
	ChannelIDs         []string
	CreatedAt          time.Time
	UpdatedAt          time.Time
}

// Alert represents a stateful alert event.
type Alert struct {
	ID                   string
	OrganizationID       string
	WorkspaceID          *string
	ResourceID           *string
	PolicyID             string
	MonitorID            string
	MonitorName          string
	State                string
	Severity             string
	Title                string
	Description          string
	DedupKey             string
	ConsecutiveFailures  int
	ConsecutiveSuccesses int
	OpenedAt             *time.Time
	ResolvedAt           *time.Time
	AcknowledgedAt       *time.Time
	AcknowledgedBy       *string
	LastNotifiedAt       *time.Time
	NotificationCount    int
	CreatedAt            time.Time
}

// EvaluateResult is the outcome of evaluating a probe result against policies.
type EvaluateResult struct {
	AlertID     string
	OldState    string
	NewState    string
	MonitorID   string
	MonitorName string
	Severity    string
	Title       string
	ChannelIDs  []string
}

// Notification carries the payload sent to notification channels.
type Notification struct {
	AlertID   string    `json:"alert_id"`
	Title     string    `json:"title"`
	State     string    `json:"state"`
	Severity  string    `json:"severity"`
	Monitor   string    `json:"monitor"`
	MonitorID string    `json:"monitor_id"`
	FiredAt   time.Time `json:"fired_at"`
}

// AlertRepository is the persistence port for alert state.
type AlertRepository interface {
	ListActivePolicies(ctx context.Context, monitorID string) ([]AlertPolicy, error)
	FindByDedup(ctx context.Context, dedupKey string) (Alert, error)
	UpsertAlert(ctx context.Context, alert *Alert) error
	ListFiring(ctx context.Context) ([]Alert, error)
	RecordNotification(ctx context.Context, alertID string) error
}

// ChannelRepository is the persistence port for notification channels.
type ChannelRepository interface {
	ListByIDs(ctx context.Context, ids []string) ([]NotificationChannel, error)
}
